File: common/src/main/java/org/valkyrienskies/clockwork/content/kinetics/resistor/RedstoneResistorBlock.java
package org.valkyrienskies.clockwork.content.kinetics.resistor;

import com.simibubi.create.content.kinetics.RotationPropagator;
import com.simibubi.create.content.kinetics.base.AbstractEncasedShaftBlock;
import com.simibubi.create.content.kinetics.base.KineticBlockEntity;
import com.simibubi.create.content.kinetics.transmission.SplitShaftBlockEntity;
import com.simibubi.create.foundation.block.IBE;
import com.simibubi.create.foundation.utility.Iterate;
import net.minecraft.core.BlockPos;
import net.minecraft.core.Direction;
import net.minecraft.server.level.ServerLevel;
import net.minecraft.world.item.context.BlockPlaceContext;
import net.minecraft.world.level.BlockGetter;
import net.minecraft.world.level.Level;
import net.minecraft.world.level.block.Block;
import net.minecraft.world.level.block.entity.BlockEntity;
import net.minecraft.world.level.block.entity.BlockEntityType;
import net.minecraft.world.level.block.state.BlockState;
import net.minecraft.world.level.block.state.StateDefinition;
import net.minecraft.world.level.block.state.properties.BlockStateProperties;
import net.minecraft.world.level.block.state.properties.BooleanProperty;
import net.minecraft.world.ticks.TickPriority;
import org.valkyrienskies.clockwork.ClockWorkBlockEntities;

import java.util.Random;

public class RedstoneResistorBlock extends AbstractEncasedShaftBlock implements IBE<SplitShaftBlockEntity> {

    public static final BooleanProperty POWERED = BlockStateProperties.POWERED;

    public RedstoneResistorBlock(Properties properties) {
        super(properties);
        registerDefaultState(defaultBlockState().setValue(POWERED, false));
    }

    @Override
    protected void createBlockStateDefinition(StateDefinition.Builder<Block, BlockState> builder) {
        builder.add(POWERED);
        super.createBlockStateDefinition(builder);
    }

    @Override
    public BlockState getStateForPlacement(BlockPlaceContext context) {
        return super.getStateForPlacement(context).setValue(POWERED,
                context.getLevel().hasNeighborSignal(context.getClickedPos()));
    }

    @Override
    public void neighborChanged(BlockState state, Level worldIn, BlockPos pos, Block blockIn, BlockPos fromPos,
                                boolean isMoving) {
        if (worldIn.isClientSide)
            return;

        boolean previouslyPowered = state.getValue(POWERED);
        if (previouslyPowered != worldIn.hasNeighborSignal(pos)) {
            detachKinetics(worldIn, pos, true);
            worldIn.setBlock(pos, state.cycle(POWERED), 2);
        }
    }

    @Override
    public Class<SplitShaftBlockEntity> getBlockEntityClass() {
        return SplitShaftBlockEntity.class;
    }

    @Override
    public BlockEntityType<? extends SplitShaftBlockEntity> getBlockEntityType() {
        return ClockWorkBlockEntities.REDSTONE_RESISTOR.get();
    }

    public void detachKinetics(Level worldIn, BlockPos pos, boolean reAttachNextTick) {
        BlockEntity te = worldIn.getBlockEntity(pos);
        if (te == null || !(te instanceof KineticBlockEntity))
            return;
        RotationPropagator.handleRemoved(worldIn, pos, (KineticBlockEntity) te);

        // Re-attach next tick
        if (reAttachNextTick)
            worldIn.scheduleTick(pos, this, 0, TickPriority.EXTREMELY_HIGH);
    }

    @Override
    public void tick(BlockState state, ServerLevel worldIn, BlockPos pos, Random random) {
        BlockEntity te = worldIn.getBlockEntity(pos);
        if (te == null || !(te instanceof KineticBlockEntity))
            return;
        KineticBlockEntity kte = (KineticBlockEntity) te;
        RotationPropagator.handleAdded(worldIn, pos, kte);
    }
}

File: common/src/main/kotlin/org/valkyrienskies/clockwork/content/kinetics/resistor/RedstoneResistorBlock.kt
package org.valkyrienskies.clockwork.content.kinetics.resistor

import com.simibubi.create.content.kinetics.RotationPropagator
import com.simibubi.create.content.kinetics.base.AbstractEncasedShaftBlock
import com.simibubi.create.content.kinetics.base.KineticBlockEntity
import com.simibubi.create.content.kinetics.transmission.SplitShaftBlockEntity
import com.simibubi.create.foundation.block.IBE
import net.minecraft.core.BlockPos
import net.minecraft.server.level.ServerLevel
import net.minecraft.util.RandomSource
import net.minecraft.world.item.context.BlockPlaceContext
import net.minecraft.world.level.Level
import net.minecraft.world.level.block.Block
import net.minecraft.world.level.block.entity.BlockEntityType
import net.minecraft.world.level.block.state.BlockState
import net.minecraft.world.level.block.state.StateDefinition
import net.minecraft.world.level.block.state.properties.BlockStateProperties
import net.minecraft.world.ticks.TickPriority
import org.valkyrienskies.clockwork.ClockworkBlockEntities

class RedstoneResistorBlock(properties: Properties) :
    AbstractEncasedShaftBlock(properties),
    IBE<SplitShaftBlockEntity> {
    init {
        registerDefaultState(defaultBlockState().setValue(POWERED, false))
    }

    override fun createBlockStateDefinition(builder: StateDefinition.Builder<Block, BlockState>) {
        builder.add(POWERED)
        super.createBlockStateDefinition(builder)
    }

    override fun getStateForPlacement(context: BlockPlaceContext): BlockState? {
        return super.getStateForPlacement(context)!!.setValue(
            POWERED,
            context.level.hasNeighborSignal(context.clickedPos)
        )
    }

    override fun neighborChanged(
        state: BlockState,
        worldIn: Level,
        pos: BlockPos,
        blockIn: Block,
        fromPos: BlockPos,
        isMoving: Boolean
    ) {
        if (worldIn.isClientSide) return
        val previouslyPowered = state.getValue(POWERED)
        if (previouslyPowered != worldIn.hasNeighborSignal(pos)) {
            detachKinetics(worldIn, pos, true)
            worldIn.setBlock(pos, state.cycle(POWERED), 2)
        }
    }

    override fun getBlockEntityClass(): Class<SplitShaftBlockEntity> {
        return SplitShaftBlockEntity::class.java
    }

    override fun getBlockEntityType(): BlockEntityType<out SplitShaftBlockEntity> {
        return ClockworkBlockEntities.REDSTONE_RESISTOR.get()
    }

    fun detachKinetics(worldIn: Level, pos: BlockPos?, reAttachNextTick: Boolean) {
        val te = worldIn.getBlockEntity(pos)
        if (te == null || te !is KineticBlockEntity) return
        RotationPropagator.handleRemoved(worldIn, pos, te as KineticBlockEntity?)

        // Re-attach next tick
        if (reAttachNextTick) worldIn.scheduleTick(pos, this, 0, TickPriority.EXTREMELY_HIGH)
    }

    override fun tick(state: BlockState, level: ServerLevel, pos: BlockPos, random: RandomSource) {
        val te = level.getBlockEntity(pos)
        if (te == null || te !is KineticBlockEntity) return
        RotationPropagator.handleAdded(level, pos, te)
    }

    companion object {
        val POWERED = BlockStateProperties.POWERED
    }
}
